File: packages/js/src/platform/services.ts
import type { PlatformClient } from './client'
import { LizardError, TimeoutError } from '../errors'

export interface Service {
  id: string
  name: string
  projectId: string
  status: 'none' | 'running' | 'crashed' | 'stopped'
  deployStatus: 'idle' | 'building' | 'deploying' | 'restarting' | 'failed' | 'deleting'
  domain?: string
  region?: string
  sourceType?: 'github' | 'upload'
  repoUrl?: string
  branch?: string
  startCommand?: string
  buildCommand?: string
  containerPort?: number
}

export interface CreateServiceOpts {
  projectId: string
  name: string
  sourceType?: 'github' | 'upload'
  repoUrl?: string
  branch?: string
  startCommand?: string
  buildCommand?: string
  containerPort?: number
  region?: string
  /** Set to 0 for worker mode (no HTTP listener). */
  port?: number
}

export interface ScaleOpts {
  replicas?: number
  cpuMillis?: number
  memoryMi?: number
  storageMi?: number
}

export interface LogLine {
  level: 'info' | 'error' | 'warn' | 'debug'
  message: string
  ts: number
  service?: string
  replica?: string
}

export interface DeployEvent {
  event: 'log' | 'done' | 'error' | 'deployed' | 'failed' | 'deploying'
  line?: string
  message?: string
  status?: string
  url?: string | null
}

/**
 * Handle for a running deploy — stream its logs or await completion.
 *
 * @example
 * ```ts
 * const deploy = await lizard.services.upload({ projectId, source: fs.readFileSync('app.tar.gz') })
 * for await (const line of deploy.logs()) console.log(line)
 * const result = await deploy.wait()
 * console.log('deployed to', result.url)
 * ```
 */
export class DeployHandle {
  private readonly _serviceId: string
  private readonly _buildId: string | undefined
  private readonly _client: PlatformClient

  constructor(client: PlatformClient, serviceId: string, buildId?: string) {
    this._client = client
    this._serviceId = serviceId
    this._buildId = buildId
  }

  get serviceId(): string { return this._serviceId }

  /** Stream build + runtime log lines until the deploy finishes. */
  async *logs(): AsyncGenerator<string> {
    const path = `/api/apps/${this._serviceId}/logs?limit=1000`
    const res = await fetch(`${this._client.config.apiUrl}${path}`, {
      headers: this._client.config.headers,
    })
    if (!res.ok) return
    const body = await res.json() as { logs?: string[] } | string[]
    const lines = Array.isArray(body) ? body : (body.logs ?? [])
    for (const l of lines) yield typeof l === 'string' ? l : JSON.stringify(l)
  }

  /**
   * Poll until the deploy reaches a terminal state.
   * @returns `{ url, status }` on success; throws `LizardError` on failure.
   */
  async wait(opts?: { timeoutMs?: number; pollMs?: number }): Promise<{ url: string | null; status: string }> {
    const deadline = Date.now() + (opts?.timeoutMs ?? 10 * 60_000)
    const pollMs = opts?.pollMs ?? 3000
    while (Date.now() < deadline) {
      const svc = await this._client.get<Service>(`/api/apps/${this._serviceId}`)
      if (svc.deployStatus === 'idle') {
        if (svc.status === 'running') return { url: svc.domain ? `https://${svc.domain}` : null, status: 'running' }
        if (svc.status === 'crashed') throw new LizardError(`Service crashed after deploy`)
      }
      if (svc.deployStatus === 'failed') throw new LizardError(`Deploy failed`)
      await new Promise(r => setTimeout(r, pollMs))
    }
    throw new TimeoutError(`Deploy did not complete within ${opts?.timeoutMs ?? 600_000}ms`)
  }
}

export class ServicesAPI {
  constructor(private readonly client: PlatformClient) {}

  /** List all services in a project. */
  list(opts: { projectId: string }): Promise<Service[]> {
    return this.client.get(`/api/projects/${opts.projectId}/apps`)
  }

  /** Get a service by ID. */
  get(id: string): Promise<Service> {
    return this.client.get(`/api/apps/${id}`)
  }

  /**
   * Create a service and start a deploy from a git repo.
   * Returns a DeployHandle to track progress.
   */
  async deploy(opts: CreateServiceOpts & { waitForDeploy?: boolean }): Promise<DeployHandle> {
    const body: Record<string, unknown> = {
      name: opts.name,
      sourceType: opts.sourceType ?? 'github',
      repoUrl: opts.repoUrl,
      branch: opts.branch ?? 'main',
      startCommand: opts.startCommand,
      buildCommand: opts.buildCommand,
      containerPort: opts.port ?? opts.containerPort,
      region: opts.region,
    }
    const svc = await this.client.post<Service>(`/api/projects/${opts.projectId}/apps`, body)
    return new DeployHandle(this.client, svc.id)
  }

  /**
   * Upload a tarball and deploy it.
   *
   * @param opts.source - A `Buffer`, `Blob`, or `Uint8Array` of a `.tar.gz` file.
   */
  async upload(opts: {
    projectId: string
    name: string
    source: Blob | string
    startCommand?: string
    buildCommand?: string
    port?: number
    region?: string
  }): Promise<DeployHandle> {
    const form = new FormData()
    const blob = opts.source instanceof Blob
      ? opts.source
      : new Blob([opts.source as string], { type: 'application/gzip' })
    form.append('file', blob, 'source.tar.gz')
    form.append('name', opts.name)
    if (opts.startCommand) form.append('startCommand', opts.startCommand)
    if (opts.buildCommand) form.append('buildCommand', opts.buildCommand)
    if (opts.port !== undefined) form.append('containerPort', String(opts.port))
    if (opts.region) form.append('region', opts.region)

    const result = await this.client.postForm<{ id: string; buildId?: string }>(
      `/api/projects/${opts.projectId}/apps/upload`,
      form,
    )
    return new DeployHandle(this.client, result.id, result.buildId)
  }

  /** Trigger a redeploy (rebuild from current source). */
  async redeploy(id: string): Promise<DeployHandle> {
    await this.client.post(`/api/apps/${id}/redeploy`, {})
    return new DeployHandle(this.client, id)
  }

  /** Restart a service without rebuilding. */
  restart(id: string): Promise<void> {
    return this.client.post(`/api/apps/${id}/restart`, {})
  }

  /** Scale a service. */
  scale(id: string, opts: ScaleOpts): Promise<void> {
    return this.client.post(`/api/apps/${id}/scale`, opts)
  }

  /**
   * Get recent log lines. For a live tail, use the WebSocket API instead.
   *
   * @param opts.limit - Max lines to return (default 200, max 1000).
   */
  async logs(id: string, opts?: { limit?: number; since?: string }): Promise<LogLine[]> {
    const qs = new URLSearchParams()
    if (opts?.limit) qs.set('limit', String(opts.limit))
    if (opts?.since) qs.set('since', opts.since)
    const result = await this.client.get<{ logs: LogLine[] } | LogLine[]>(`/api/apps/${id}/logs?${qs}`)
    return Array.isArray(result) ? result : (result.logs ?? [])
  }

  /** Execute a command inside the running service container. */
  exec(id: string, cmd: string, opts?: { timeoutMs?: number }): Promise<{ stdout: string; stderr: string; exitCode: number }> {
    return this.client.post(`/api/apps/${id}/exec`, { cmd, timeoutMs: opts?.timeoutMs })
  }

  /** Update service configuration. */
  update(id: string, opts: Partial<Pick<Service, 'name' | 'startCommand' | 'buildCommand' | 'containerPort' | 'repoUrl' | 'branch' | 'sourceType'>>): Promise<Service> {
    return this.client.patch(`/api/apps/${id}`, opts)
  }

  /** Delete a service. */
  delete(id: string): Promise<void> {
    return this.client.delete(`/api/apps/${id}`)
  }
}

File: packages/python/lizard/platform/services.py
from __future__ import annotations
import time
from dataclasses import dataclass, field
from typing import TYPE_CHECKING, Iterator

if TYPE_CHECKING:
    from .client import PlatformClient

from ..errors import LizardError, TimeoutError as LizardTimeoutError


@dataclass
class Service:
    id: str
    name: str
    project_id: str
    status: str
    deploy_status: str
    domain: str | None = None
    region: str | None = None
    source_type: str | None = None
    repo_url: str | None = None
    branch: str | None = None

    @classmethod
    def _from_dict(cls, d: dict) -> "Service":
        return cls(
            id=d["id"],
            name=d["name"],
            project_id=d.get("projectId", ""),
            status=d.get("status", "none"),
            deploy_status=d.get("deployStatus", "idle"),
            domain=d.get("domain"),
            region=d.get("region"),
            source_type=d.get("sourceType"),
            repo_url=d.get("repoUrl"),
            branch=d.get("branch"),
        )


@dataclass
class LogLine:
    level: str
    message: str
    ts: int
    service: str | None = None
    replica: str | None = None

    @classmethod
    def _from_dict(cls, d: dict) -> "LogLine":
        return cls(
            level=d.get("level", "info"),
            message=d.get("message", ""),
            ts=int(d.get("ts", 0)),
            service=d.get("service"),
            replica=d.get("replica"),
        )


class DeployHandle:
    """
    Handle for a running deploy. Poll for completion or stream logs.

    Example::

        deploy = lizard.services.upload(project_id=pid, name="api", source=data)
        for line in deploy.logs():
            print(line)
        result = deploy.wait()
        print("Deployed to", result["url"])
    """

    def __init__(self, client: "PlatformClient", service_id: str) -> None:
        self._client = client
        self.service_id = service_id

    def logs(self, limit: int = 200) -> list[str]:
        """Return recent log lines."""
        try:
            result = self._client.get(f"/api/apps/{self.service_id}/logs?limit={limit}")
            if isinstance(result, list):
                return [l if isinstance(l, str) else str(l) for l in result]
            return [l if isinstance(l, str) else str(l) for l in result.get("logs", [])]
        except Exception:
            return []

    def wait(self, *, timeout_ms: int = 10 * 60_000, poll_ms: int = 3000) -> dict:
        """
        Block until the deploy reaches a terminal state.

        Returns ``{"url": str | None, "status": str}`` on success.
        Raises ``LizardError`` on failure or ``TimeoutError`` on timeout.
        """
        deadline = time.time() + timeout_ms / 1000
        while time.time() < deadline:
            svc_dict = self._client.get(f"/api/apps/{self.service_id}")
            deploy_status = svc_dict.get("deployStatus", "")
            status = svc_dict.get("status", "")
            if deploy_status == "idle":
                if status == "running":
                    domain = svc_dict.get("domain")
                    return {"url": f"https://{domain}" if domain else None, "status": "running"}
                if status == "crashed":
                    raise LizardError("Service crashed after deploy")
            if deploy_status == "failed":
                raise LizardError("Deploy failed")
            time.sleep(poll_ms / 1000)
        raise LizardTimeoutError(f"Deploy did not complete within {timeout_ms}ms")


class ServicesAPI:
    def __init__(self, client: "PlatformClient") -> None:
        self._client = client

    def list(self, *, project_id: str) -> list[Service]:
        """List all services in a project."""
        return [Service._from_dict(s) for s in self._client.get(f"/api/projects/{project_id}/apps")]

    def get(self, id: str) -> Service:
        """Get a service by ID."""
        return Service._from_dict(self._client.get(f"/api/apps/{id}"))

    def deploy(
        self,
        *,
        project_id: str,
        name: str,
        repo_url: str | None = None,
        branch: str = "main",
        source_type: str = "github",
        start_command: str | None = None,
        build_command: str | None = None,
        port: int | None = None,
        region: str | None = None,
    ) -> DeployHandle:
        """Create a service and kick off a git-source deploy."""
        body = {
            "name": name,
            "sourceType": source_type,
            "repoUrl": repo_url,
            "branch": branch,
            "startCommand": start_command,
            "buildCommand": build_command,
            "containerPort": port,
            "region": region,
        }
        svc = self._client.post(f"/api/projects/{project_id}/apps", {k: v for k, v in body.items() if v is not None})
        return DeployHandle(self._client, svc["id"])

    def upload(
        self,
        *,
        project_id: str,
        name: str,
        source: bytes,
        start_command: str | None = None,
        build_command: str | None = None,
        port: int | None = None,
        region: str | None = None,
    ) -> DeployHandle:
        """Upload a `.tar.gz` tarball and deploy it."""
        extra: dict[str, str] = {"name": name}
        if start_command:
            extra["startCommand"] = start_command
        if build_command:
            extra["buildCommand"] = build_command
        if port is not None:
            extra["containerPort"] = str(port)
        if region:
            extra["region"] = region
        result = self._client.post_file(
            f"/api/projects/{project_id}/apps/upload",
            "source.tar.gz",
            source,
            extra,
        )
        return DeployHandle(self._client, result["id"])

    def redeploy(self, id: str) -> DeployHandle:
        """Trigger a redeploy (rebuild from current source)."""
        self._client.post(f"/api/apps/{id}/redeploy")
        return DeployHandle(self._client, id)

    def restart(self, id: str) -> None:
        """Restart a service without rebuilding."""
        self._client.post(f"/api/apps/{id}/restart")

    def scale(
        self,
        id: str,
        *,
        replicas: int | None = None,
        cpu_millis: int | None = None,
        memory_mi: int | None = None,
        storage_mi: int | None = None,
    ) -> None:
        """Scale a service."""
        body = {k: v for k, v in {
            "replicas": replicas,
            "cpuMillis": cpu_millis,
            "memoryMi": memory_mi,
            "storageMi": storage_mi,
        }.items() if v is not None}
        self._client.post(f"/api/apps/{id}/scale", body)

    def logs(self, id: str, *, limit: int = 200) -> list[LogLine]:
        """Get recent log lines."""
        result = self._client.get(f"/api/apps/{id}/logs?limit={limit}")
        rows = result if isinstance(result, list) else result.get("logs", [])
        return [LogLine._from_dict(l) if isinstance(l, dict) else LogLine(level="info", message=str(l), ts=0)
                for l in rows]

    def exec_(self, id: str, cmd: str, *, timeout_ms: int | None = None) -> dict:
        """Execute a command inside the running service container."""
        body = {"cmd": cmd}
        if timeout_ms is not None:
            body["timeoutMs"] = str(timeout_ms)
        return self._client.post(f"/api/apps/{id}/exec", body)

    def delete(self, id: str) -> None:
        """Delete a service."""
        self._client.delete(f"/api/apps/{id}")
